# "Used cars from eBay" data exploration

In this project, we'll work with a dataset of used cars from eBay Kleinanzeigen, a classifieds section of the German eBay website.

The dataset was originally scraped and uploaded to Kaggle by user orgesleka.
The original dataset isn't available on Kaggle anymore, but you can find it here.

We've made a few modifications from the original dataset:

-  We sampled 50,000 data points from the full dataset, to ensure your code runs quickly in our hosted environment
-  We dirtied the dataset a bit to more closely resemble what you would expect from a scraped dataset (the version uploaded to Kaggle was cleaned to be easier to work with)

The data dictionary provided with data is as follows:

-  dateCrawled - When this ad was first crawled. All field-values are taken from this date.
-  name'- Name of the car.
-  seller - Whether the seller is private or a dealer.
-  offerType - The type of listing
-  price - The price on the ad to sell the car.
-  abtest - Whether the listing is included in an A/B test.
-  vehicleType - The vehicle Type.
-  yearOfRegistration - The year in which the car was first registered.
-  gearbox - The transmission type.
-  powerPS - The power of the car in PS.
-  model - The car model name.
-  kilometer - How many kilometers the car has driven.
-  monthOfRegistration - The month in which the car was first registered.
-  fuelType - What type of fuel the car uses.
-  brand - The brand of the car.
-  notRepairedDamage - If the car has a damage which is not yet repaired.
-  dateCreated - The date on which the eBay listing was created.
-  nrOfPictures - The number of pictures in the ad.
-  postalCode - The postal code for the location of the vehicle.
-  lastSeenOnline - When the crawler saw this ad last online.

Now, lets open and explore the dataframe.

In [2]:
import pandas as pd
autos = pd.read_csv('autos.csv', encoding = 'Latin-1')
autos.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 20 columns):
 #   Column               Non-Null Count  Dtype 
---  ------               --------------  ----- 
 0   dateCrawled          50000 non-null  object
 1   name                 50000 non-null  object
 2   seller               50000 non-null  object
 3   offerType            50000 non-null  object
 4   price                50000 non-null  object
 5   abtest               50000 non-null  object
 6   vehicleType          44905 non-null  object
 7   yearOfRegistration   50000 non-null  int64 
 8   gearbox              47320 non-null  object
 9   powerPS              50000 non-null  int64 
 10  model                47242 non-null  object
 11  odometer             50000 non-null  object
 12  monthOfRegistration  50000 non-null  int64 
 13  fuelType             45518 non-null  object
 14  brand                50000 non-null  object
 15  notRepairedDamage    40171 non-null  object
 16  date

In our dataset, the following observations are made:
-  The dataset contains 20 columns and most of which are strings. 
-  Some columns have null values, but none have more than ~20% null values.
-  Column titles are mix of high and low letters with characters that requires cleaning.

Now lets start the data cleaning.

## Initial Data Exploration and Cleaning

We'll start by exploring the data to find obvious areas where we can clean the data.

In [3]:
autos.columns

Index(['dateCrawled', 'name', 'seller', 'offerType', 'price', 'abtest',
       'vehicleType', 'yearOfRegistration', 'gearbox', 'powerPS', 'model',
       'odometer', 'monthOfRegistration', 'fuelType', 'brand',
       'notRepairedDamage', 'dateCreated', 'nrOfPictures', 'postalCode',
       'lastSeen'],
      dtype='object')

We'll make a few changes here:

Change the columns from camelcase to snakecase.
Change a few wordings to more accurately describe the columns.

In [4]:
new_columns = ['datecrawled', 'name', 'seller', 'offertype', 'price', 'abtest',
       'vehicletype', 'registration_year', 'gearbox', 'powerps', 'model',
       'odometer', 'registration_month', 'fueltype', 'brand',
       'unrepaired_damage', 'ad_created', 'nofpictures', 'postalcode',
       'lastseen']
autos.columns = new_columns
autos.head(3)

datecrawled                                        name  seller  \
0  2016-03-26 17:47:46            Peugeot_807_160_NAVTECH_ON_BOARD  privat   
1  2016-04-04 13:38:56  BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik  privat   
2  2016-03-26 18:57:24                  Volkswagen_Golf_1.6_United  privat   

  offertype   price   abtest vehicletype  registration_year    gearbox  \
0   Angebot  $5,000  control         bus               2004    manuell   
1   Angebot  $8,500  control   limousine               1997  automatik   
2   Angebot  $8,990     test   limousine               2009    manuell   

   powerps   model   odometer  registration_month fueltype       brand  \
0      158  andere  150,000km                   3      lpg     peugeot   
1      286     7er  150,000km                   6   benzin         bmw   
2      102    golf   70,000km                   7   benzin  volkswagen   

  unrepaired_damage           ad_created  nofpictures  postalcode  \
0              nein  2016-03-26 00:00:00            0       79588   
1              nein  2016-04-04 00:00:00            0       71034   
2              nein  2016-03-26 00:00:00            0       35394   

              lastseen  
0  2016-04-06 06:45:54  
1  2016-04-06 14:45:08  
2  2016-04-06 20:15:37

In [5]:
autos.describe(include = 'all')

datecrawled         name  seller offertype  price abtest  \
count                 50000        50000   50000     50000  50000  50000   
unique                48213        38754       2         2   2357      2   
top     2016-03-30 19:48:02  Ford_Fiesta  privat   Angebot     $0   test   
freq                      3           78   49999     49999   1421  25756   
mean                    NaN          NaN     NaN       NaN    NaN    NaN   
std                     NaN          NaN     NaN       NaN    NaN    NaN   
min                     NaN          NaN     NaN       NaN    NaN    NaN   
25%                     NaN          NaN     NaN       NaN    NaN    NaN   
50%                     NaN          NaN     NaN       NaN    NaN    NaN   
75%                     NaN          NaN     NaN       NaN    NaN    NaN   
max                     NaN          NaN     NaN       NaN    NaN    NaN   

       vehicletype  registration_year  gearbox       powerps  model  \
count        44905       50000.000000    47320  50000.000000  47242   
unique           8                NaN        2           NaN    245   
top      limousine                NaN  manuell           NaN   golf   
freq         12859                NaN    36993           NaN   4024   
mean           NaN        2005.073280      NaN    116.355920    NaN   
std            NaN         105.712813      NaN    209.216627    NaN   
min            NaN        1000.000000      NaN      0.000000    NaN   
25%            NaN        1999.000000      NaN     70.000000    NaN   
50%            NaN        2003.000000      NaN    105.000000    NaN   
75%            NaN        2008.000000      NaN    150.000000    NaN   
max            NaN        9999.000000      NaN  17700.000000    NaN   

         odometer  registration_month fueltype       brand unrepaired_damage  \
count       50000        50000.000000    45518       50000             40171   
unique         13                 NaN        7          40                 2   
top     150,000km                 NaN   benzin  volkswagen              nein   
freq        32424                 NaN    30107       10687             35232   
mean          NaN            5.723360      NaN         NaN               NaN   
std           NaN            3.711984      NaN         NaN               NaN   
min           NaN            0.000000      NaN         NaN               NaN   
25%           NaN            3.000000      NaN         NaN               NaN   
50%           NaN            6.000000      NaN         NaN               NaN   
75%           NaN            9.000000      NaN         NaN               NaN   
max           NaN           12.000000      NaN         NaN               NaN   

                 ad_created  nofpictures    postalcode             lastseen  
count                 50000      50000.0  50000.000000                50000  
unique                   76          NaN           NaN                39481  
top     2016-04-03 00:00:00          NaN           NaN  2016-04-07 06:17:27  
freq                   1946          NaN           NaN                    8  
mean                    NaN          0.0  50813.627300                  NaN  
std                     NaN          0.0  25779.747957                  NaN  
min                     NaN          0.0   1067.000000                  NaN  
25%                     NaN          0.0  30451.000000                  NaN  
50%                     NaN          0.0  49577.000000                  NaN  
75%                     NaN          0.0  71540.000000                  NaN  
max                     NaN          0.0  99998.000000                  NaN

In [6]:
autos["price"].value_counts()

$0          1421
$500         781
$1,500       734
$2,500       643
$1,000       639
            ... 
$277           1
$117           1
$37,700        1
$4,510         1
$128,000       1
Name: price, Length: 2357, dtype: int64

In [7]:
autos["odometer"].value_counts()

150,000km    32424
125,000km     5170
100,000km     2169
90,000km      1757
80,000km      1436
70,000km      1230
60,000km      1164
50,000km      1027
5,000km        967
40,000km       819
30,000km       789
20,000km       784
10,000km       264
Name: odometer, dtype: int64

So, its clear that:
-  Any columns that have mostly one value that are candidates to be dropped
-  Any columns that need more investigation.
-  Any examples of numeric data stored as text that needs to be cleaned.

In [8]:
autos["price"] = autos["price"].str.replace("$","").str.replace(",","").astype(int)
autos["price"].head()


0    5000
1    8500
2    8990
3    4350
4    1350
Name: price, dtype: int64

In [9]:
autos["odometer"] = autos["odometer"].str.replace("km", "").str.replace(",", "").astype(int)

autos.rename({"odometer": "odometer_km"}, inplace = True, axis = 1)
autos["odometer_km"].head()

0    150000
1    150000
2     70000
3     70000
4    150000
Name: odometer_km, dtype: int64

In [10]:
autos["nofpictures"].value_counts()

0    50000
Name: nofpictures, dtype: int64

In [11]:
autos = autos.drop(["seller", "offertype","nofpictures"], axis = 1)
autos.head()

datecrawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   price   abtest vehicletype  registration_year    gearbox  powerps   model  \
0   5000  control         bus               2004    manuell      158  andere   
1   8500  control   limousine               1997  automatik      286     7er   
2   8990     test   limousine               2009    manuell      102    golf   
3   4350  control  kleinwagen               2007  automatik       71  fortwo   
4   1350     test       kombi               2003    manuell        0   focus   

   odometer_km  registration_month fueltype       brand unrepaired_damage  \
0       150000                   3      lpg     peugeot              nein   
1       150000                   6   benzin         bmw              nein   
2        70000                   7   benzin  volkswagen              nein   
3        70000                   6   benzin       smart              nein   
4       150000                   7   benzin        ford              nein   

            ad_created  postalcode             lastseen  
0  2016-03-26 00:00:00       79588  2016-04-06 06:45:54  
1  2016-04-04 00:00:00       71034  2016-04-06 14:45:08  
2  2016-03-26 00:00:00       35394  2016-04-06 20:15:37  
3  2016-03-12 00:00:00       33729  2016-03-15 03:16:28  
4  2016-04-01 00:00:00       39218  2016-04-01 14:38:50

## Exploring Odometer and Price

Let's continue exploring the data, specifically looking for data that doesn't look right. We'll start by analyzing the 'odometer_km' and 'price' columns.

In [12]:
print(autos["price"].unique().shape)
print('\n')
print(autos["price"].value_counts().sort_index(ascending = False).head(15))
print('\n')
print(autos["price"].value_counts().sort_index(ascending = True).head(15))

(2357,)


99999999    1
27322222    1
12345678    3
11111111    2
10000000    1
3890000     1
1300000     1
1234566     1
999999      2
999990      1
350000      1
345000      1
299000      1
295000      1
265000      1
Name: price, dtype: int64


0     1421
1      156
2        3
3        1
5        2
8        1
9        1
10       7
11       2
12       3
13       2
14       1
15       2
17       3
18       1
Name: price, dtype: int64


In the price variable, there are total 2357 unique numbers, of which 1421 are have zero value and 14 have price avobe ~1 million dollars. Since our dataset includes old cars only, its unlikely that car price would reach that high value so those values needs to be dropped.

In [13]:
autos = autos[autos["price"].between(1, 350000)]
print(autos["price"].unique().shape)
print(autos.describe())

(2346,)
               price  registration_year       powerps    odometer_km  \
count   48565.000000       48565.000000  48565.000000   48565.000000   
mean     5888.935591        2004.755421    117.197158  125770.101925   
std      9059.854754          88.643887    200.649618   39788.636804   
min         1.000000        1000.000000      0.000000    5000.000000   
25%      1200.000000        1999.000000     71.000000  125000.000000   
50%      3000.000000        2004.000000    107.000000  150000.000000   
75%      7490.000000        2008.000000    150.000000  150000.000000   
max    350000.000000        9999.000000  17700.000000  150000.000000   

       registration_month    postalcode  
count        48565.000000  48565.000000  
mean             5.782251  50975.745207  
std              3.685595  25746.968398  
min              0.000000   1067.000000  
25%              3.000000  30657.000000  
50%              6.000000  49716.000000  
75%              9.000000  71665.000000  
max    

After removing outliers in the price values, ~2.8% or 1,435 rows were dropped from the original datasets. Now, lets move on to 'odometer'.

In [14]:
autos['odometer_km'].value_counts().sort_index(ascending = True)

5000        836
10000       253
20000       762
30000       780
40000       815
50000      1012
60000      1155
70000      1217
80000      1415
90000      1734
100000     2115
125000     5057
150000    31414
Name: odometer_km, dtype: int64

For the 'odometer_km', all parameters were rounded and there are no significant outliers. Next, lets see registration_year as it showed min parameter of 1000. 

In [15]:
print(autos["registration_year"].value_counts().sort_index(ascending = False).head(15))
print('\n')
autos["registration_year"].value_counts().sort_index(ascending = True).head(15)

9999       3
9000       1
8888       1
6200       1
5911       1
5000       4
4800       1
4500       1
4100       1
2800       1
2019       2
2018     470
2017    1392
2016    1220
2015     392
Name: registration_year, dtype: int64




1000    1
1001    1
1111    1
1800    2
1910    5
1927    1
1929    1
1931    1
1934    2
1937    4
1938    1
1939    1
1941    2
1943    1
1948    1
Name: registration_year, dtype: int64

There are 17 rows shows registration year of greater than 2800 and 5 rows are less than 1800. Those outliers volume is relatively small, so lets drop those columns as well.

In [16]:
autos = autos[autos['registration_year'].between(1910,2020)]
print(autos.describe())

               price  registration_year       powerps    odometer_km  \
count   48545.000000       48545.000000  48545.000000   48545.000000   
mean     5888.431764        2003.461510    117.156123  125819.136883   
std      9059.345265           7.566196    200.385362   39723.379171   
min         1.000000        1910.000000      0.000000    5000.000000   
25%      1200.000000        1999.000000     71.000000  125000.000000   
50%      3000.000000        2004.000000    107.000000  150000.000000   
75%      7490.000000        2008.000000    150.000000  150000.000000   
max    350000.000000        2019.000000  17700.000000  150000.000000   

       registration_month    postalcode  
count        48545.000000  48545.000000  
mean             5.783603  50975.125121  
std              3.684938  25748.944248  
min              0.000000   1067.000000  
25%              3.000000  30657.000000  
50%              6.000000  49716.000000  
75%              9.000000  71665.000000  
max            

## Exploring date columns

There are 5 columns that should represent date values. Some of these columns were created by the crawler, some came from the website itself. We can differentiate by referring to the data dictionary:

- `date_crawled`: added by the crawler
- `last_seen`: added by the crawler
- `ad_created`: from the website
- `registration_month`: from the website
- `registration_year`: from the website

And date columns come from crawled are recognized as string by panda and lets start by converting their types.

In [17]:
print(autos[['datecrawled','ad_created','lastseen']][0:5])

           datecrawled           ad_created             lastseen
0  2016-03-26 17:47:46  2016-03-26 00:00:00  2016-04-06 06:45:54
1  2016-04-04 13:38:56  2016-04-04 00:00:00  2016-04-06 14:45:08
2  2016-03-26 18:57:24  2016-03-26 00:00:00  2016-04-06 20:15:37
3  2016-03-12 16:58:10  2016-03-12 00:00:00  2016-03-15 03:16:28
4  2016-04-01 14:38:50  2016-04-01 00:00:00  2016-04-01 14:38:50


In [18]:
(autos['datecrawled']
                  .str[:10]
                  .value_counts(normalize = True, dropna = False)
                  .sort_index()
                  ) 


2016-03-05    0.025317
2016-03-06    0.014028
2016-03-07    0.036028
2016-03-08    0.033309
2016-03-09    0.033103
2016-03-10    0.032197
2016-03-11    0.032568
2016-03-12    0.036914
2016-03-13    0.015676
2016-03-14    0.036564
2016-03-15    0.034277
2016-03-16    0.029622
2016-03-17    0.031600
2016-03-18    0.012916
2016-03-19    0.034792
2016-03-20    0.037882
2016-03-21    0.037388
2016-03-22    0.033000
2016-03-23    0.032197
2016-03-24    0.029354
2016-03-25    0.031620
2016-03-26    0.032218
2016-03-27    0.031105
2016-03-28    0.034854
2016-03-29    0.034051
2016-03-30    0.033701
2016-03-31    0.031826
2016-04-01    0.033659
2016-04-02    0.035472
2016-04-03    0.038603
2016-04-04    0.036482
2016-04-05    0.013101
2016-04-06    0.003172
2016-04-07    0.001401
Name: datecrawled, dtype: float64

In [19]:
(autos['ad_created']
 .str[:10]
 .value_counts(normalize = True, dropna = False)
 .sort_index()
)

2015-06-11    0.000021
2015-08-10    0.000021
2015-09-09    0.000021
2015-11-10    0.000021
2015-12-05    0.000021
                ...   
2016-04-03    0.038851
2016-04-04    0.036852
2016-04-05    0.011824
2016-04-06    0.003255
2016-04-07    0.001257
Name: ad_created, Length: 76, dtype: float64

In [20]:
(autos['lastseen']
 .str[:10]
 .value_counts(normalize = True, dropna = False)
 .sort_index()
)

2016-03-05    0.001071
2016-03-06    0.004326
2016-03-07    0.005397
2016-03-08    0.007416
2016-03-09    0.009599
2016-03-10    0.010671
2016-03-11    0.012380
2016-03-12    0.023772
2016-03-13    0.008899
2016-03-14    0.012607
2016-03-15    0.015882
2016-03-16    0.016459
2016-03-17    0.028077
2016-03-18    0.007354
2016-03-19    0.015820
2016-03-20    0.020661
2016-03-21    0.020641
2016-03-22    0.021382
2016-03-23    0.018539
2016-03-24    0.019775
2016-03-25    0.019219
2016-03-26    0.016809
2016-03-27    0.015656
2016-03-28    0.020867
2016-03-29    0.022350
2016-03-30    0.024761
2016-03-31    0.023751
2016-04-01    0.022783
2016-04-02    0.024925
2016-04-03    0.025214
2016-04-04    0.024472
2016-04-05    0.124730
2016-04-06    0.221794
2016-04-07    0.131939
Name: lastseen, dtype: float64

The crawler recorded the date it last saw any listing, which allows us to determine on what day a listing was removed, presumably because the car was sold.
As we see above, data were crawled within 34 days and lastseen and datacrawled were identical if we excluding the timeframe. For the date the ad created, its shows that 10 monhts of dataset were crawled. And finally, The last three days contain a disproportionate amount of 'last seen' values. Given that these are 6-10x the values from the previous days, it's unlikely that there was a massive spike in sales, and more likely that these values are to do with the crawling period ending and don't indicate car sales.

In [21]:
autos["registration_year"].describe()

count    48545.000000
mean      2003.461510
std          7.566196
min       1910.000000
25%       1999.000000
50%       2004.000000
75%       2008.000000
max       2019.000000
Name: registration_year, dtype: float64

## Exploring Price by Brand

In [22]:
autos['brand'].value_counts(normalize = True)

volkswagen        0.212813
opel              0.108641
bmw               0.108641
mercedes_benz     0.095787
audi              0.085858
ford              0.069667
renault           0.047894
peugeot           0.029457
fiat              0.025996
seat              0.018931
skoda             0.016068
nissan            0.015264
mazda             0.015223
smart             0.014296
citroen           0.014111
toyota            0.012586
hyundai           0.009950
sonstige_autos    0.009641
volvo             0.009043
mini              0.008611
mitsubishi        0.008178
honda             0.007993
kia               0.007107
alfa_romeo        0.006612
porsche           0.005912
suzuki            0.005891
chevrolet         0.005665
chrysler          0.003481
dacia             0.002657
daihatsu          0.002513
jeep              0.002204
subaru            0.002101
land_rover        0.002039
saab              0.001627
daewoo            0.001566
jaguar            0.001524
trabant           0.001380
r

German manufacturers represent four out of the top five brands, almost 50% of the overall listings. Volkswagen is by far the most popular brand, with approximately double the cars for sale of the next two brands combined.

There are lots of brands that don't have a significant percentage of listings, so we will limit our analysis to brands representing more than 5% of total listings.

In [23]:
brand_dist = autos['brand'].value_counts(normalize = True)
brand_5above = brand_dist[brand_dist > 0.05].index   #Remember that Series.value_counts() produces a series with index labels, so you can use Series.index attribute to access the labels, should you wish.


In [24]:
mean_price = {}

for b in brand_5above:
        brand = autos[autos['brand'] == b]
        mean_p = brand['price'].mean()
        mean_price[b] = int(mean_p)
        
print(mean_price)

{'volkswagen': 5333, 'opel': 2941, 'bmw': 8261, 'mercedes_benz': 8526, 'audi': 9212, 'ford': 3728}


Within 5% or more common brands:
-  aud, bmw and mercedes benz are more pricy as mean prices are 8000 or higher,
-  volswagen's price ranges somewhere in middle, ~5000,
-  and opel and ford are the cheapest, ranging ~3000
    

## Exploring Mileage

In [25]:
mean_mileage = {}

for b in brand_5above:
    brandm = autos[autos['brand'] == b]
    mileage = brandm['odometer_km'].mean()
    mean_mileage[b] = int(mileage)

print(mean_mileage)

{'volkswagen': 128955, 'opel': 129452, 'bmw': 132682, 'mercedes_benz': 130848, 'audi': 129492, 'ford': 124349}


In [26]:
df_mil = pd.Series(mean_mileage) #converting to data series
df_prc = pd.Series(mean_price).sort_values(ascending = False)

df_comb = pd.DataFrame(df_prc, columns = ['mean_price'])
df_comb['mean_mil'] = df_mil #adding second series to the dataframe
print(df_comb)

               mean_price  mean_mil
audi                 9212    129492
mercedes_benz        8526    130848
bmw                  8261    132682
volkswagen           5333    128955
ford                 3728    124349
opel                 2941    129452


After cleaning the dataset, we have left with 48,545 samples of old german car on eBay. For old car prices, we explored common brands that has 5% above distribution with the sample dataset and among them, audi, mercedes benz and bmw are the most pricy brands. However, when it comes to its mean mileage there are not much differences.

Now lets explore more that how prices ranges between different group of mileage.

In [67]:
autos['odometer_km'].value_counts()

150000    31414
125000     5057
100000     2115
90000      1734
80000      1415
70000      1217
60000      1155
50000      1012
5000        823
40000       815
30000       780
20000       762
10000       246
Name: odometer_km, dtype: int64

In [75]:
# Tobecontineud

## Price differences between non-damage and with damaged cars

Now lets see How much cheaper are cars with damage than their non-damaged counterparts.

In [27]:
autos.describe(include = 'all')
autos['unrepaired_damage'].value_counts()

nein    34771
ja       4689
Name: unrepaired_damage, dtype: int64

In [45]:
dtype = autos['unrepaired_damage'].value_counts().index
damage_price = {}

for t in dtype:
    dty = autos[autos['unrepaired_damage'] == t]
    price = dty['price'].mean()
    damage_price[t] = price
    
print(damage_price)


{'nein': 7086.102355411119, 'ja': 2221.887609298358}


From the result we can see that non damaged cars' price ranges around ~7000, while damaged cars priced 
almost 3 times lower, around 2000 dollars.

## Conclusion

After cleaning the dataset, we have left with 48,545 samples of old german car on eBay. For old car prices, we explored common brands that has 5% above distribution with the sample dataset and among them, audi, mercedes benz and bmw are the most pricy brands. However, when it comes to its mean mileage there are not much differences.

For further polishing reference: https://github.com/Lippins/Ebay_car_sales/blob/main/notebook.ipynb